webhooks/external_account_config.go

```
@@ -0,0 +1,81 @@
+	ServiceAccountImpersonationURL string `json:"service_account_impersonation_url"`
+	// ServiceAccountImpersonationLifetimeSeconds is the number of seconds the service account impersonation
+	// token will be valid for.
+	ServiceAccountImpersonationLifetimeSeconds int `json:"service_account_impersonation_lifetime_seconds,omitempty"`
+	// CredentialSource contains the necessary information to retrieve the token itself, as well
+	// as some environmental information.
+	CredentialSource CredentialSource `json:"credential_source"`
+}
+
+// CredentialSource stores the information necessary to retrieve the credentials for the STS exchange.
+// One field amongst File, URL, and Executable should be filled, depending on the kind of credential in question.
+// The EnvironmentID should start with AWS if being used for an AWS credential.
+type CredentialSource struct {
+	File   string           `json:"file"`
+	Format CredentialFormat `json:"format"`
+}
+
+type CredentialFormat struct {
+	// Type is either "text" or "json". When not provided "text" type is assumed.
+	Type string `json:"type"`
+}
+
+func NewExternalAccountCredentials(aud, gsaEmail string) *ExternalAccountCredentials {
+	creds := &ExternalAccountCredentials{
+		Type:             "external_account",
+		Audience:         aud,
+		SubjectTokenType: "urn:ietf:params:oauth:token-type:jwt",
+		TokenURL:         "https://sts.googleapis.com/v1/token",
+		CredentialSource: CredentialSource{
+			File:   filepath.Join(K8sSATokenMountPath, K8sSATokenName),
+			Format: CredentialFormat{Type: "text"},
+		},
+		ServiceAccountImpersonationURL: fmt.Sprintf("https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/%s:generateAccessToken", gsaEmail),
+	}
+
+	return creds
+}
+
+func (e *ExternalAccountCredentials) String() string {
+	b, err := json.MarshalIndent(e, "", "  ")
+	if err != nil {
+		log.Printf("could not marshal ")
+		return ""
+	}
```

Should we return error here instead of returning empty string?? Then, we can log this error in the caller side (mutator?).

And I think `log.Printf` is not better because other components use `logr`.　If we really log the warning here, we would wiring up logger instance to this struct.

WDYT?

---

Oh yes, good catch. I did not mean to use a different logger, or leave this error case unresolved. I'll take another pass.

---

I've pushed a new commit that fixes this. PTAL 🙇